### PR TITLE
Validation cleanup part 7

### PR DESCRIPTION
**Pull request by thockin**
More cleanups, wordsmithing, document conventions.

### CODE REVIEW

File: docs/devel/api-conventions.md [hunk 1]
```diff
@@ -787,6 +788,35 @@ There are two primary protocols in use today:
...
+API objects are validated upon receipt by the apiserver.  Validation errors are
+flagged and returned to the caller in a `Failure` status with `reason` set to
+`Invalid`.  In order to facilitate consistent error messages, we ask that
+validation logic adheres to the following guidelines whenever possible (though
+exceptional cases will exist).
+
+* Be as precise as possible.
+* Telling users what they CAN do is more useful than telling them what they
+  CANNOT do.
+* When asserting a requirement in the positive, use "must".  Examples: "must be
```
Comment (caesarxuchao) [inline docs/devel/api-conventions.md]: A `formatting requirement`? It should be consistent with the next point.

File: pkg/api/validation/validation.go [hunk 2]
```diff
@@ -1111,22 +1224,6 @@ func validateHostNetwork(hostNetwork bool, containers []api.Container, fldPath *
 	return allErrors
 }
 
-// validateImagePullSecrets checks to make sure the pull secrets are well
-// formed.  Right now, we only expect name to be set (it's the only field).  If
-// this ever changes and someone decides to set those fields, we'd like to
-// know.
-func validateImagePullSecrets(imagePullSecrets []api.LocalObjectReference, fldPath *field.Path) field.ErrorList {
```
Comment (caesarxuchao) [inline pkg/api/validation/validation.go]: Why removing this? It this field getting deprecated?

File: pkg/api/validation/validation.go [hunk 3]
```diff
@@ -1547,12 +1646,12 @@ func ValidateNodeUpdate(node, oldNode *api.Node) field.ErrorList {
...
 	allErrs := field.ErrorList{}
 	if !validation.IsQualifiedName(value) {
-		return append(allErrs, field.Invalid(fldPath, value, "resource typename: "+qualifiedNameErrorMsg))
+		return append(allErrs, field.Invalid(fldPath, "", qualifiedNameErrorMsg))
 	}
 
 	if len(strings.Split(value, "/")) == 1 {
 		if !api.IsStandardResourceName(value) {
-			return append(allErrs, field.Invalid(fldPath, value, "is neither a standard resource type nor is fully qualified"))
+			return append(allErrs, field.Invalid(fldPath, "", "must be a standard resource type or fully qualified"))
```
Comment (caesarxuchao) [inline pkg/api/validation/validation.go]: Why remove the wrong `value`?

Comment (caesarxuchao) [inline pkg/api/validation/validation.go]: I'll continue my pass later today.

File: docs/devel/api-conventions.md (hunk 1 above)
Comment (thockin) [inline docs/devel/api-conventions.md]: I specifically split the negative case into formatting and behavioral, but in the positive sense there is no difference in words.

File: pkg/api/validation/validation.go (hunk 2 above)
Comment (thockin) [inline pkg/api/validation/validation.go]: This is why we do code review - merge artifact. :)

File: pkg/api/validation/validation.go (hunk 3 above)
Comment (thockin) [inline pkg/api/validation/validation.go]: My mistake, I thought it was the value, but the error is the key.  But 'value' here is the key :)

**Comment by thockin** [comment]
rebased and fixed, thanks

File: pkg/api/validation/validation.go [hunk 4]
```diff
@@ -1784,7 +1901,7 @@ func ValidateResourceRequirements(requirements *api.ResourceRequirements, fldPat
 				limitValue = quantity.MilliValue()
 			}
 			if limitValue < requestValue {
-				allErrs = append(allErrs, field.Invalid(fldPath, quantity.String(), "limit cannot be smaller than request"))
+				allErrs = append(allErrs, field.Invalid(fldPath, quantity.String(), "must be greater than or equal to `request`"))
```
Comment (caesarxuchao) [inline pkg/api/validation/validation.go]: A nit. There is no field named `request`. so how about "must be greater than or equal to the counterpart in \`requests\`"?

File: pkg/api/validation/validation.go [hunk 5]
```diff
@@ -2053,18 +2170,17 @@ func ValidateSecurityContext(sc *api.SecurityContext, fldPath *field.Path) field
...
 	}
 	if opts.LimitBytes != nil && *opts.LimitBytes < 1 {
-		allErrs = append(allErrs, field.Invalid(field.NewPath("limitBytes"), *opts.LimitBytes, "limitBytes must be a positive integer or nil"))
+		allErrs = append(allErrs, field.Invalid(field.NewPath("limitBytes"), *opts.LimitBytes, "must be greater than 0"))
 	}
 	switch {
 	case opts.SinceSeconds != nil && opts.SinceTime != nil:
-		allErrs = append(allErrs, field.Invalid(field.NewPath("sinceSeconds"), *opts.SinceSeconds, "only one of sinceTime or sinceSeconds can be provided"))
-		allErrs = append(allErrs, field.Invalid(field.NewPath("sinceTime"), *opts.SinceTime, "only one of sinceTime or sinceSeconds can be provided"))
+		allErrs = append(allErrs, field.Invalid(field.NewPath(""), *opts.SinceSeconds, "at most one of `sinceTime` or `sinceSeconds` may be specified"))
```
Comment (caesarxuchao) [inline pkg/api/validation/validation.go]: Should be a Forbidden.

File: pkg/api/validation/validation_test.go [hunk 6]
```diff
@@ -919,7 +919,7 @@ func TestValidateEnv(t *testing.T) {
 					},
 				},
 			}},
-			expectedError: "valueFrom.fieldRef.fieldPath: unsupported value 'status.phase', Details: supported values: metadata.name, metadata.namespace, status.podIP",
+			expectedError: `valueFrom.fieldRef.fieldPath: Unsupported value: "status.phase": supported values: metadata.name, metadata.namespace, status.podIP`,
```
Comment (caesarxuchao) [inline pkg/api/validation/validation_test.go]: The multiple colons in a single sentence still makes me frown. Is this following some tradition?

File: pkg/api/validation/validation_test.go [hunk 7]
```diff
@@ -4172,7 +4172,7 @@ func TestValidateSecurityContext(t *testing.T) {
 	}
 	for k, v := range successCases {
 		if errs := ValidateSecurityContext(v.sc, field.NewPath("field")); len(errs) != 0 {
-			t.Errorf("Expected success for %s, got %v", k, errs)
+			t.Errorf("[%s Expected success, got %v", k, errs)
```
Comment (caesarxuchao) [inline pkg/api/validation/validation_test.go]: missing `]`

File: pkg/apis/extensions/validation/validation.go [hunk 8]
```diff
@@ -42,16 +42,16 @@ func ValidateHorizontalPodAutoscalerName(name string, prefix bool) (bool, string
...
-		allErrs = append(allErrs, field.Invalid(fldPath.Child("maxReplicas"), autoscaler.MaxReplicas, `must be greater than or equal to 1`))
+		allErrs = append(allErrs, field.Invalid(fldPath.Child("maxReplicas"), autoscaler.MaxReplicas, "must be greater than 0"))
 	}
 	if autoscaler.MinReplicas != nil && autoscaler.MaxReplicas < *autoscaler.MinReplicas {
-		allErrs = append(allErrs, field.Invalid(fldPath.Child("maxReplicas"), autoscaler.MaxReplicas, `must be greater than or equal to minReplicas`))
+		allErrs = append(allErrs, field.Invalid(fldPath.Child("maxReplicas"), autoscaler.MaxReplicas, "must be greater than or equal to `minReplicas`"))
 	}
 	if autoscaler.CPUUtilization != nil && autoscaler.CPUUtilization.TargetPercentage < 1 {
-		allErrs = append(allErrs, field.Invalid(fldPath.Child("cpuUtilization", "targetPercentage"), autoscaler.CPUUtilization.TargetPercentage, `must be greater than or equal to 1`))
+		allErrs = append(allErrs, field.Invalid(fldPath.Child("cpuUtilization", "targetPercentage"), autoscaler.CPUUtilization.TargetPercentage, "must be greater than or equal to 1"))
```
Comment (caesarxuchao) [inline pkg/apis/extensions/validation/validation.go]: Nit. Consider changing to "must be greater than 0".

File: pkg/apis/extensions/validation/validation.go [hunk 9]
```diff
@@ -570,14 +570,14 @@ func ValidateLabelSelectorRequirement(sr extensions.LabelSelectorRequirement, fl
 	switch sr.Operator {
 	case extensions.LabelSelectorOpIn, extensions.LabelSelectorOpNotIn:
 		if len(sr.Values) == 0 {
-			allErrs = append(allErrs, field.Invalid(fldPath.Child("values"), sr.Values, "must be non-empty when operator is In or NotIn"))
+			allErrs = append(allErrs, field.Invalid(fldPath.Child("values"), sr.Values, "must be specified when `operator` is 'In' or 'NotIn'"))
 		}
 	case extensions.LabelSelectorOpExists, extensions.LabelSelectorOpDoesNotExist:
 		if len(sr.Values) > 0 {
-			allErrs = append(allErrs, field.Invalid(fldPath.Child("values"), sr.Values, "must be empty when operator is Exists or DoesNotExist"))
+			allErrs = append(allErrs, field.Invalid(fldPath.Child("values"), sr.Values, "must not be specified when `operator` is 'Exists' or 'DoesNotExist'"))
```
Comment (caesarxuchao) [inline pkg/apis/extensions/validation/validation.go]: Change to `Forbidden` and `may not be`?

File: pkg/apis/extensions/validation/validation.go [hunk 10]
```diff
@@ -570,14 +570,14 @@ func ValidateLabelSelectorRequirement(sr extensions.LabelSelectorRequirement, fl
 	switch sr.Operator {
 	case extensions.LabelSelectorOpIn, extensions.LabelSelectorOpNotIn:
 		if len(sr.Values) == 0 {
-			allErrs = append(allErrs, field.Invalid(fldPath.Child("values"), sr.Values, "must be non-empty when operator is In or NotIn"))
+			allErrs = append(allErrs, field.Invalid(fldPath.Child("values"), sr.Values, "must be specified when `operator` is 'In' or 'NotIn'"))
```
Comment (caesarxuchao) [inline pkg/apis/extensions/validation/validation.go]: Forbidden?

File: pkg/api/validation/validation.go (hunk 4 above)
Comment (thockin) [inline pkg/api/validation/validation.go]: how about just removing the backticks.  It's clear to me.

File: pkg/apis/extensions/validation/validation_test.go [hunk 11]
```diff
@@ -871,7 +871,7 @@ func TestValidateJob(t *testing.T) {
 				Template: validPodTemplateSpec,
 			},
 		},
-		"spec.template.metadata.labels: invalid value 'map[y:z]', Details: selector does not match template": {
+		"spec.template.metadata.labels: Invalid value: {\"y\":\"z\"}: `selector` does not match template": {
```
Comment (caesarxuchao) [inline pkg/apis/extensions/validation/validation_test.go]: Could you add  \`labels\` at the end?

File: pkg/api/validation/validation.go (hunk 5 above)
Comment (thockin) [inline pkg/api/validation/validation.go]: done

File: pkg/api/validation/validation_test.go (hunk 6 above)
Comment (thockin) [inline pkg/api/validation/validation_test.go]: It's pretty go-style - add details: at the front.

File: pkg/api/validation/validation_test.go (hunk 7 above)
Comment (thockin) [inline pkg/api/validation/validation_test.go]: done

File: pkg/apis/extensions/validation/validation.go (hunk 8 above)
Comment (thockin) [inline pkg/apis/extensions/validation/validation.go]: done

**Comment by caesarxuchao** [comment]
Just made a complete pass. I'll take a look at your replies tomorrow. Thank you for the cleanup!

File: pkg/apis/extensions/validation/validation.go (hunk 9 above)
Comment (thockin) [inline pkg/apis/extensions/validation/validation.go]: Done

File: pkg/apis/extensions/validation/validation.go (hunk 10 above)
Comment (thockin) [inline pkg/apis/extensions/validation/validation.go]: Required

File: pkg/apis/extensions/validation/validation_test.go (hunk 11 above)
Comment (thockin) [inline pkg/apis/extensions/validation/validation_test.go]: done

Comment (thockin-cc) [inline pkg/apis/extensions/validation/validation_test.go]: Pushed

**Comment by thockin** [comment]
@k8s-bot test this please

@k8s-bot unit test this please

**Comment by caesarxuchao** [comment]
Unit/Integration test looks like a real problem:
```
message:The Service "invalid-large-service-name" is invalid.
metadata.name: Invalid value: "invalid-large-service-name": must be a DNS 952 label (at most 24 characters, matching regex [a-z]([-a-z0-9]*[a-z0-9])?): e.g. "my-name"
has not:metadata.name: invalid value
983 ./hack/test-cmd.sh
```

**Comment by caesarxuchao** [comment]
e2e failure is also real, need to change error message:
```
 <string>: Service "nodeport-collision-2" is invalid: spec.ports[0].nodePort: Invalid value: 31604: provided port is already allocated
to equal
    <string>: Service "nodeport-collision-2" is invalid: spec.ports[0].nodePort: invalid value '31604', Details: provided port is already allocated
```

**Comment by caesarxuchao** [comment]
Changes LGTM. Feel free to apply the label when you fix the tests.

File: pkg/api/validation/validation.go [hunk 12]
```diff
@@ -1548,12 +1665,12 @@ func ValidateNodeUpdate(node, oldNode *api.Node) field.ErrorList {
 func validateResourceName(value string, fldPath *field.Path) field.ErrorList {
 	allErrs := field.ErrorList{}
 	if !validation.IsQualifiedName(value) {
-		return append(allErrs, field.Invalid(fldPath, value, "resource typename: "+qualifiedNameErrorMsg))
+		return append(allErrs, field.Invalid(fldPath, "", qualifiedNameErrorMsg))
```
Comment (caesarxuchao) [inline pkg/api/validation/validation.go]: Adding back the value here?

**Comment by thockin** [comment]
@k8s-bot test this

**Comment by thockin** [comment]
@k8s-bbot unit test this please

**Comment by thockin** [comment]
@k8s-bot unit test this please

**Comment by thockin** [comment]
@k8s-bot unit test this please

**Comment by thockin** [comment]
@ixdy is the bot not running tests?

**Comment by thockin** [comment]
Error from git checkout: reference is not a tree: 612af72f9d9fac7e13b0c8a535b04fed96262124
Sigh

**Comment by thockin** [comment]
@k8s-bot unit test this please

**Comment by caesarxuchao** [comment]
Seems to be a flake
```
=== RUN TestWatchEtcdError
panic: test timed out after 5m0s
```

**Comment by caesarxuchao** [comment]
@k8s-bot unit test this

**Pull request merged**